src/atlas/interpolation/method/sphericalvector/SphericalVector.cc

```diff
@@ -1,132 +1,57 @@
 /*
- * (C) Crown Copyright 2023 Met Office
+ * (C) Crown Copyright 2024 Met Office
  *
  * This software is licensed under the terms of the Apache Licence Version 2.0
  * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
  */
 
-#include "atlas/library/defines.h"
 #include "atlas/interpolation/method/sphericalvector/SphericalVector.h"
 
-#include <cmath>
-#include <tuple>
-#include <type_traits>
-#include <utility>
-
-#include "eckit/config/LocalConfiguration.h"
+// I don't yet fully trust NVHPC. Turning off OpenMP until we can figure out why
+// it's failing CI.
+#ifdef __NVCOMPILER
+#warning turning off OpenMP for nvhpc build
+#undef atlas_omp_parallel_for
+#define atlas_omp_parallel_for for
+#endif
 
 #include "atlas/array/ArrayView.h"
-#include "atlas/array/helpers/ArrayForEach.h"
-#include "atlas/array/Range.h"
 #include "atlas/field/Field.h"
 #include "atlas/field/FieldSet.h"
 #include "atlas/interpolation/Cache.h"
 #include "atlas/interpolation/Interpolation.h"
 #include "atlas/interpolation/method/MethodFactory.h"
+#include "atlas/interpolation/method/sphericalvector/ComplexMatrixMultiply.h"
+#include "atlas/interpolation/method/sphericalvector/Types.h"
+#include "atlas/option/Options.h"
 #include "atlas/parallel/omp/omp.h"
 #include "atlas/runtime/Exception.h"
 #include "atlas/runtime/Trace.h"
 #include "atlas/util/Constants.h"
 #include "atlas/util/Geometry.h"
+#include "eckit/config/LocalConfiguration.h"
 
 namespace atlas {
 namespace interpolation {
 namespace method {
 
-namespace  {
-MethodBuilder<SphericalVector> __builder("spherical-vector");
-}
-
-#if ATLAS_HAVE_EIGEN
-
-// A bug exists in intel versions < intel/2022.2 with OpenMP
-// Intel OneAPI version 2022.2 corresponds to Intel classic (icpc) version 2021.6
-#if defined(__INTEL_COMPILER) && defined(__INTEL_COMPILER_UPDATE)
-#if (__INTEL_COMPILER <= 2021) && (__INTEL_COMPILER_UPDATE < 6)
-#warning Disabling OpenMP to prevent internal compiler error for intel-classic version < 2021.6 (intel-oneapi/2022.2)
-#undef atlas_omp_parallel_for
-#define atlas_omp_parallel_for for
-#endif
```

Comment:
How come you could drop this?

Reply:
I think the problem arose when supplying a template parameter functor to loop body. I've since separated out the use-cases into their own loops, as @pmaciel suggested. No longer triggers the error. 🙂